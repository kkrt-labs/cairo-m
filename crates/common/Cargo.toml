[package]
name = "cairo-m-common"
version = "0.1.0"
edition.workspace = true

[dependencies]
num-traits.workspace = true
paste = "1.0"
serde.workspace = true
serde_json.workspace = true
thiserror.workspace = true
smallvec.workspace = true

# Using workspace dependency for stwo
stwo-prover.workspace = true
