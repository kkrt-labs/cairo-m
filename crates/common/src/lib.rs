pub mod instruction;
pub mod opcode;
pub mod program;

pub use instruction::Instruction;
pub use opcode::Opcode;
pub use program::{Program, ProgramMetadata};
