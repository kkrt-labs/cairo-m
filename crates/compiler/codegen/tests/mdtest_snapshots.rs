//! Codegen snapshot tests for mdtest cases.
//! This file automatically generates codegen snapshots for all Cairo-M code in the mdtest directory,
//! providing comprehensive coverage of real-world examples through the entire compilation pipeline.

mod common;

use cairo_m_compiler_codegen::CodeGenerator;
use cairo_m_compiler_mir::generate_mir;
use cairo_m_test_utils::{mdtest::MdTestRunner, mdtest_path};
use common::{create_test_crate, TestDatabase};

#[test]
fn test_mdtest_codegen_snapshots() {
    use insta::{assert_snapshot, glob, with_settings};

    // Use glob! to discover all markdown files in mdtest directory
    glob!(mdtest_path().to_str().unwrap(), "**/*.md", |path| {
        let db = TestDatabase::default();

        let runner = MdTestRunner::new("CASM", |source, name| {
            let crate_id = create_test_crate(&db, source, name, "mdtest");

            // First generate MIR
            let mir_module = match generate_mir(&db, crate_id) {
                Ok(module) => module,
                Err(diagnostics) => {
                    return Err(format!("MIR generation failed: {:#?}", diagnostics));
                }
            };

            // Then generate CASM code from MIR
            let mut generator = CodeGenerator::new();
            match generator.generate_module(&mir_module) {
                Ok(_) => Ok(generator.debug_instructions()),
                Err(e) => Err(format!("Code generation failed: {:#?}", e)),
            }
        })
        .with_parent_dir(true);

        let snapshots = runner.run_file(path);

        for snapshot in snapshots {
            with_settings!({
                description => format!("Codegen snapshot for mdtest: {}", snapshot.name).as_str(),
                omit_expression => true,
                snapshot_suffix => snapshot.suffix,
                prepend_module_to_snapshot => false,
            }, {
                assert_snapshot!(snapshot.content);
            });
        }
    });
}
