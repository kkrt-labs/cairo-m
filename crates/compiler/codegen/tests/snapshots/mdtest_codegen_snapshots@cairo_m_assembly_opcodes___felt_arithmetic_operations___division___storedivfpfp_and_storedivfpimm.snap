---
source: crates/compiler/codegen/tests/mdtest_snapshots.rs
description: "Codegen snapshot for mdtest: Cairo-M Assembly Opcodes - Felt Arithmetic Operations - Division - StoreDivFpFp and StoreDivFpImm"
input_file: mdtest/06-internals/01-opcodes.md
---
Source:
fn test_division() -> felt {
    let a = 20;
    let b = 4;

    // StoreDivFpFp: [fp + dst_off] = [fp + src0_off] / [fp + src1_off]
    let div1 = a / b;

    // StoreDivFpImm: [fp + dst_off] = [fp + src_off] / imm
    let div2 = a / 2;

    return div1 + div2;
}
============================================================
Generated CASM:
test_division:
test_division:
test_division_0:
   0: 9 20 0 _             // [fp + 0] = 20
   1: 9 4 1 _              // [fp + 1] = 4
   2: 3 0 1 2              // [fp + 2] = [fp + 0] op [fp + 1]
   3: 7 0 2 3              // [fp + 3] = [fp + 0] op 2
   4: 0 2 3 -3             // [fp + -3] = [fp + 2] op [fp + 3]
   5: 11 _ _ _             // return
